content/blog/real-time-smart-city-traffic-monitoring-using-microservices-based-streaming-architecture-part-2.md

```diff
@@ -0,0 +1,331 @@
+
+To be honest, I’m only scratching the surface of what this kind of smart road could do to make a commuter’s life both easier and safer while saving money for the city. But how would we build a prototype of such a system without it being hugely complicated and expensive?
+
+## How to Build a Rule-Engine Based CEP System
+
+So we now have a proper, concrete target in mind. It turns out that if we decide to base our system around reading our data from a Kafka-style stream (i.e., persistent, scalable, and high performance), then we will naturally end up with a pretty cool, modern CEP microservice.
+
+The important point here is not to show how to build a super complicated enterprise architecture, but rather that by making some simple technology choices and building our demo in a reasonable way, we **naturally end up with** this elegant, modern, and **simple architecture**.
+
+For simplicity’s sake, I have decided to implement my prototype on a MapR Sandbox<sup>*</sup>. This is because it will include the stream messaging system, MapR Event Store, which I can use through the Kafka 0.9 API with very little configuration and know it will work the same on a production MapR 5.1+ cluster.
+
+_<sup>*</sup>Use [The Development Environment for HPE Ezmeral Data Fabric](https://docs.datafabric.hpe.com/62/MapRContainerDevelopers/MapRContainerDevelopersOverview.html)._
+
+Finally, it should be noted that an Apache Kafka cluster could be used as well with the same design and code, just with some additional work to get it up and running.
+
+*High-Level Architecture View*
+
+![](/img/picture2.png)
+
+As shown in the diagram above, the flow is to have sensor data get aggregated to a producer gateway, which will forward the data to the stream with a topic named “data.” The data will be in JSON format so it is easy to manipulate, human readable, and easily sent to Elasticsearch as-is for monitoring with a Kibana dashboard.
+
+The consumer side will have two tasks: to read the data from the stream and host an instance of a KieSession where the rule engine can apply rules on the facts as they are added to it.
+
+Rules are edited in the Workbench GUI, a Java webapp which can be run on a Java application server such as <a target='\_blank'  href='http://wildfly.org/'>WildFly</a> or <a target='\_blank'  href='http://tomcat.apache.org/'>Tomcat</a>.
+
+Rules are fetched from the workbench by the consumer application using the appropriate methods provided by the Drools framework, which are entirely based on Maven Repository.
+
+We can now look at the proposed system in more detail in the following sections.
+
+**List of Technologies Used**
+
+The technology we’re going to use is as follows:
+
+* MapR Sandbox 5.2
```

Comment:
You might want to double check the bullets you're using here... something looks off:

![image](https://user-images.githubusercontent.com/77496261/137515593-fdf066d5-9107-4f6d-a9fe-35dc6931b302.png)

Reply:
done